File: java/src/test/java/no/bouvet/workshop/solid/solution/OpenClosedCsvReader.java
package no.bouvet.workshop.solid.solution;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class OpenClosedCsvReader {

    public void parse(File file) throws IOException {
        visitStart(file);
        try {
            try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    String[] country = line.split(",");
                    visitLine(format(country));
                }
            }
        } catch (IOException e) {
            handle(file, e);
        }
        visitEnd(file);
    }

    protected void visitStart(File file) {
        System.out.println("Starting to parse: " + file);
    }

    protected void visitEnd(File file) {
        System.out.println("Done parsing: " + file);
    }

    protected void visitLine(String formatted) {
        System.out.println(formatted);
    }

    protected String format(String[] country) {
        return "Country [code= " + country[4] + " , name=" + country[5] + "]";
    }

    protected void handle(File file, IOException exception) throws IOException {
        throw exception;
    }
}

File: dotNet/delegation/test/Solution/OpenClosedCsvReader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace test.Solution
{
    public class OpenClosedCsvReader
    {
        public void parse(FileInfo file)
        {
            visitStart(file);
            try
            {
                using (var reader = file.OpenText())
                {
                    string line;
                    while ((line = reader.ReadLine()) != null)
                    {
                        string[] country = line.Split(',');
                        visitLine(format(country));
                    }
                }
            }
            catch(IOException e)
            {
                handle(file, e);
            }
            visitEnd(file);
        }

        protected virtual void visitStart(FileInfo file)
        {
            Console.WriteLine("Starting to parse: " + file);
        }

        protected virtual void visitEnd(FileInfo file)
        {
            Console.WriteLine("Done parsing: " + file);
        }

        protected virtual void visitLine(String formatted)
        {
            Console.WriteLine(formatted);
        }

        protected virtual string format(string[] country)
        {
            return "Country [code= " + country[4] + " , name=" + country[5] + "]";
        }

        protected virtual void handle(FileInfo file, IOException exception)
        {
            throw exception;
        }
    }
}
